lib/Sema/CSSimplify.cpp

```diff
@@ -3225,6 +3225,23 @@ bool ConstraintSystem::repairFailures(
     if (!isValueOfRawRepresentable(expectedType, rawReprType))
       return false;
 
+    auto valueTy = expectedType->lookThroughSingleOptionalType();
+    auto unwrappedRawReprTy = rawReprType->lookThroughSingleOptionalType();
+    auto memberLoc = getConstraintLocator(
+        locator.withPathElement(ConstraintLocator::ApplyFunction));
+    auto memberTy = createTypeVariable(memberLoc, TVO_CanBindToNoEscape);
+    addValueMemberConstraint(
+        MetatypeType::get(unwrappedRawReprTy, getASTContext()),
+        DeclNameRef(DeclName(getASTContext(), DeclBaseName::createConstructor(),
+                             {getASTContext().Id_rawValue})),
+        memberTy, DC, FunctionRefKind::SingleApply, {},
+        getConstraintLocator(
+            locator.withPathElement(ConstraintLocator::ConstructorMember)));
+    addConstraint(ConstraintKind::ApplicableFunction,
+                  FunctionType::get({AnyFunctionType::Param(valueTy)},
+                                    unwrappedRawReprTy),
+                  memberTy, memberLoc);
```

Comment:
This looks indeed pretty good so far, now you just need to associated result type of the member with expected type to make sure it picks the right overload, only nuance is that we don't know whether result type of `init` invocation would be optional or not, might need to to use a disjunction here...

Reply:
I think that's what this constraint does?

Reply:
Since result of the call is stored into `unwrappedRawReprTy` you can related that with `expected` type.